File: myFeed/Services/BackgroundService.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using DryIocAttributes;
using myFeed.Interfaces;
using myFeed.Platform;

namespace myFeed.Services
{
    [Reuse(ReuseType.Singleton)]
    [ExportEx(typeof(IBackgroundService))]
    public sealed class BackgroundService : IBackgroundService
    {
        private readonly INotificationService _notificationService;
        private readonly IFeedStoreService _feedStoreService;
        private readonly ICategoryManager _categoryManager;
        private readonly ISettingManager _settingManager;

        public BackgroundService(
            INotificationService notificationService,
            IFeedStoreService feedStoreService,
            ICategoryManager categoryManager,
            ISettingManager settingManager)
        {
            _notificationService = notificationService;
            _feedStoreService = feedStoreService;
            _categoryManager = categoryManager;
            _settingManager = settingManager;
        }
        
        public async Task<bool> CheckForUpdates(DateTime dateTime)
        {
            var categories = await _categoryManager.GetAll().ConfigureAwait(false);
            var notifyables = categories
                .SelectMany(category => category.Channels)
                .Where(channel => channel.Notify);

            var articles = await _feedStoreService.Load(notifyables).ConfigureAwait(false);
            var settings = await _settingManager.Read().ConfigureAwait(false);
            var recentArticles = articles
                .Where(article => article.PublishedDate > settings.Fetched)
                .OrderByDescending(article => article.PublishedDate)
                .Take(15).Reverse().ToList();

            await _notificationService.SendNotifications(recentArticles).ConfigureAwait(false);
            if (!recentArticles.Any()) return false;
            settings.Fetched = dateTime;
            await _settingManager.Write(settings).ConfigureAwait(false);
            return true;
        }
    }
}

File: myFeed.Tests/BackgroundService.fs
module myFeed.Tests.BackgroundService

open Xunit
open NSubstitute
open myFeed.Models
open myFeed.Interfaces
open myFeed.Services
open myFeed.Platform
open myFeed.Tests
open System.Threading.Tasks
open System.Collections.Generic
open System 

[<Fact>]
let ``should send ordered notifications for articles with greater date``() =

    let store = Substitute.For<IFeedStoreService>()
    store.Load(Arg.Any()).Returns(Task.FromResult
        ([ Article(Title="Foo", PublishedDate=DateTime.Now);
          Article(Title="Bar", PublishedDate=DateTime.MaxValue) ] :> seq<_>)) |> ignore

    let settings = Substitute.For<ISettingManager>()
    settings.Read().Returns(Settings(Fetched = DateTime.MinValue)) |> ignore

    let mutable received = null
    let notify = Substitute.For<INotificationService>()
    notify.When(fun x -> x.SendNotifications(Arg.Any()) |> ignore)
          .Do(fun x -> received <- x.Arg<List<Article>>())

    let service = produce<BackgroundService> [store; settings; notify]
    service.CheckForUpdates(DateTime.Now).Wait()
    Should.equal "Foo" received.[0].Title
    Should.equal "Bar" received.[1].Title

[<Fact>]
let ``should not send notifications for outdated old articles``() =

    let store = Substitute.For<IFeedStoreService>()
    store.Load(Arg.Any()).Returns(Task.FromResult
        ([ Article(PublishedDate=DateTime.MinValue) ] :> seq<_>)) |> ignore
        
    let settings = Substitute.For<ISettingManager>()
    settings.Read().Returns(Settings()) |> ignore

    let mutable received = null
    let notify = Substitute.For<INotificationService>()
    notify.When(fun x -> x.SendNotifications(Arg.Any()) |> ignore)
          .Do(fun x -> received <- x.Arg<List<Article>>())

    let service = produce<BackgroundService> [store; notify; settings]
    service.CheckForUpdates(DateTime.Now).Wait()
    Should.equal 0 received.Count 
    